src/key_value_store.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Seahorse library.
+
+// This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
+// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+// You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! The key-value store.
+//!
+//! This module defines [KeyValueStore], which provides interfaces to keystore resources (e.g.,
+//! the assets resource) to create, read, update, and delete data.
+
+use crate::{persistence::EncryptingResourceAdapter, KeystoreError, Ledger};
+use atomic_store::AppendLog;
+use serde::{de::DeserializeOwned, Serialize};
+use snafu::Snafu;
+use std::collections::HashMap;
+use std::hash::Hash;
+
+/// Errors happening during key-value store operations.
+#[derive(Debug, Snafu)]
+#[snafu(visibility(pub))]
+pub enum KeyValueStoreError {
+    KeyNotFound,
+    PersistenceError {
+        source: atomic_store::error::PersistenceError,
+    },
+}
+
+impl From<atomic_store::error::PersistenceError> for KeyValueStoreError {
+    fn from(source: atomic_store::error::PersistenceError) -> Self {
+        Self::PersistenceError { source }
+    }
+}
+
+impl<L: Ledger> From<KeyValueStoreError> for KeystoreError<L> {
+    fn from(source: KeyValueStoreError) -> Self {
+        KeystoreError::KeyValueStoreError { source }
+    }
+}
+
+/// A storage for key-value pairs.
+pub struct KeyValueStore<
+    K: DeserializeOwned + Eq + Hash + Serialize,
+    V: DeserializeOwned + Serialize,
+> {
+    store: AppendLog<EncryptingResourceAdapter<HashMap<K, V>>>,
```

Comment:
Why AppendLog, not RollingLog?

It's generally not advised to use `AppendLog` for entire snapshots of a `HashMap`. If you want all of history, make it an `AppendLog<EncryptingResourceAdapter<(K, Option<V>)>>`. If you want to be able to get the latest and recent entries, `RollingLog<EncryptingResourceAdapter<HashMap<K, V>>>`, which is slower to store, but should load faster. For many, but not all, use cases, `AppendLog<EncryptingResourceAdapter<(K, Option<V>)>>` would take up less space. and for all use cases it would be a much faster call to `store_resource`. But `AppendLog<EncryptingResourceAdapter<HashMap<K, V>>>` is the worst of all worlds.

Reply:
Thanks @nyospe. I think it makes sense to use `AppendLog<EncryptingResourceAdapter<(K, Option<V>)>>` rather than `RollingLog<EncryptingResourceAdapter<HashMap<K, V>>>` here since it takes less space and has a faster call to `store_resource`. I'm not sure if this is suitable for the `Assets` resource though, so would like to double-check with @jbearer.

`Assets` currently has a `commit` function as follows: 
```
pub fn commit<L: Ledger>(&mut self) -> Result<(), KeystoreError<L>> {
    Ok(self.store.commit_version()?)
}
```
where `self.store` is `KeyValueStore<AssetCode, Asset>`.

If we change the `KeyValueStore` to:
```
store: AppendLog<EncryptingResourceAdapter<(K, Option<V>)>>,
index: HashMap<K, V>,
```
then we need to provide specific key and value to the `commit_version` function which calls `store_resource((key, Some(value)))`, wheres in the current implementation we can just do `store_resource(&self.index)` to store the full in-memory index. For `Assets`, do we want to keep the `commit` function with no specific key-value pair? If so, I think we need to use `RollingLog<EncryptingResourceAdapter<HashMap<K, V>>>` for `KeyValueStore`.

Reply:
I'm confused. `commit` just calls `commit_version` on the underlying store, which is the same regardless of the data format. `store_resource` is called by `AssetEditor::save`, which already has access to the key and value.

Reply:
`commit_version` also calls `store_resource`, so we need to add `key` and `value` as the inputs of `commit_version` so that it can call `store_resource((key, Some(value)))` (similarly, `revert_version` can then call `store_resource((key, None))`). Maybe the issue is that we actually should not call `store_resource` in `commit_version` and `revert_version`? Then we don't have to change `commit` in `Assets` to take the key and value.

Reply:
Oh, yeah. I think we should not be calling store_resource in commit_version/revert_version. That makes a little sense in the RollingLog version, where the "resource" you're storing is an updated snapshot of the entire state, but doesn't really make sense in the AppendLog version where resources are individual entries.